manual/src/cmds/intf-c.etex

```diff
@@ -1474,6 +1474,25 @@ Example:
     }
 \end{verbatim}
 
+If the C function returned a \verb"void" function that is used as an \verb"extern" function in OCaml, it would be a type error.
+The C function should return the C type \verb"value".
+
+Example:
+\begin{verbatim}
+    CAMLprim value call_caml_f_ex(value closure, value arg)
+    {
+      CAMLparam2(closure, arg);
+      CAMLlocal2(res, tmp);
+      res = caml_callback_exn(closure, Val_unit);
+      if(Is_exception_result(res)) {
+        res = Extract_exception(res);
+        tmp = caml_alloc(3, 0); /* Safe to allocate: res contains valid value. */
+        ...
+      }
+      CAMLreturn(Val_unit);
+    }
+\end{verbatim}
+
```

Comment:
This is in section “Callbacks from C to OCaml”. I feel that this is not quite the right place, as we want to warn specifically about C functions called from OCaml.

I would have put this as a warning (`\paragraph{Warning:} ...`) after shortly after the introduction of the `CAMLreturn*` and `CAMLparam*` macros, in section 5.1, and phrased it similarly to what @gasche proposed in the original issue:

> clarify the documentation of the C interface in the manual to explain that CAMLreturn0 should only be used for internal FFI procedures of type void, and that CAMLreturn(Val_unit) should be used for functions that return an OCaml unit value.

Not that the obligation of returning `value`s to OCaml is already specified near the beginning of the chapter, in section 1.2:

> User primitives with arity n ≤ 5 are implemented by C functions that take n arguments of type value, and return a result of type value.

However, I am in favor of reminding it for the specific case of `CAMLreturn0` vs unit, when this macro is introduced, as the mistake is particularly easy to make.

Reply:
Yes, this sounds like the right thing to do. The original issue mentioned `caml_callback_exn` but this is a red herring, other things can probably go wrong by having garbage as a return value, such as crashing the GC.

Reply:
Yes, that is exactly what I experienced.

Reply:
There are some udpates, please help to review @gadmm and @OlivierNicole
Thanks, guys! 😄